### PR TITLE
feat(cb2-5712): Create Vehicles

### PR DESCRIPTION
**Pull request by tom-bjss**
Added ability to create vehicles

### CODE REVIEW

File: src/app/services/technical-record/technical-record.service.ts [hunk 1]
```diff
@@ -64,7 +92,38 @@ export class TechnicalRecordService {
...
+    const recordCopy = cloneDeep(newVehicleRecord);
+
+    const url = `${environment.VTM_API_URI}/vehicles`;
+    const body = {
+      msUserDetails: { msOid: user.id, msUser: user.name },
+      vin: recordCopy.vin,
+      primaryVrm: recordCopy.vrms ? recordCopy.vrms[0].vrm : '',
+      trailerId: recordCopy.trailerId ?? '',
+      techRecord: recordCopy.techRecord
+    };
```
Comment (gjulien-bjss) [inline src/app/services/technical-record/technical-record.service.ts:105]: would we ever want secondary vrms>

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.html [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        <div class="govuk-grid-column-full">
+          <app-tech-record-title [vehicleTechRecord]="vehicle || undefined" [hideActions]="true"></app-tech-record-title>
+          <app-button id="submit-record-continue" (clicked)="handleSubmit()">Submit new record</app-button>
+        </div>
+      </div>
+    </div>
+
+    <div class="govuk-grid-column-two-thirds">
+      <div class="govuk-grid-column-full">
+        <app-tech-record-summary [isEditing]="true" [vehicleTechRecord]="vehicle.techRecord[0]"></app-tech-record-summary>
```
Comment (aaron-downing) [inline src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.html:14]: Realise this isn't what you've written but after not looking at VTM for a while I had to double check for my own sanity... but I think that vehicleTechRecord is badly named for what it stores, I'd have presumed it stored the whole vehicle object, and not just a technical record object...

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import { GlobalErrorService } from '@core/components/global-error/global-error.service';
+import { Axle, VehicleTechRecordModel } from '@models/vehicle-tech-record.model';
+import { Actions, ofType } from '@ngrx/effects';
+import { Store } from '@ngrx/store';
+import { TechnicalRecordService } from '@services/technical-record/technical-record.service';
+import { createVehicleRecord, createVehicleRecordSuccess } from '@store/technical-records';
+import { TechnicalRecordServiceState } from '@store/technical-records/reducers/technical-record-service.reducer';
+import { take, Observable } from 'rxjs';
+
+@Component({
```
Comment (aaron-downing) [inline src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts:13]: Is this component representing a page that allows you to fill in details for a new record?

Comment (tom-bjss) [reply src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts:13]: The rest of the data correct

File: src/app/forms/services/dynamic-form.service.ts [hunk 4]
```diff
@@ -122,7 +122,7 @@ export class DynamicFormService {
     validators.forEach(v => control.addAsyncValidators(this.asyncValidatorMap[v.name](v.args)));
   }
 
-  static updateValidity(form: CustomFormGroup | CustomFormArray, errors: GlobalError[]) {
+  static updateValidity(form: FormGroup | FormArray, errors: GlobalError[]) {
```
Comment (gjulien-bjss) [inline src/app/forms/services/dynamic-form.service.ts:125]: why was this changed?

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts (hunk 3 above)
Comment (aaron-downing) [reply src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts:13]: HydrateNewVehicleRecordComponent seems like the name for an action/function that modifies some kind of data object more than something representing a page, just wondering if it could be named differently?

File: src/app/forms/services/dynamic-form.service.ts (hunk 4 above)
Comment (tom-bjss) [reply src/app/forms/services/dynamic-form.service.ts:125]: You were on the call when I talked about htis change 😆, so because both custom types extend the base and now im using a base group this works and it allowed me to reuse it 👍

Comment (tom-bjss) [reply src/app/forms/services/dynamic-form.service.ts:125]: As theyres nothing specific to either of those types used and the required cast was already there

File: src/app/features/tech-record/components/tech-record-summary/tech-record-summary.component.ts [hunk 5]
```diff
@@ -201,7 +201,7 @@ export class TechRecordSummaryComponent implements OnInit {
         }
       } else if (vehicleAxles && !vehicleAxleSpacings) {
         this.vehicleTechRecordCalculated.dimensions!.axleSpacing = this.generateAxleSpacing(vehicleAxles.length);
-      } else if (!vehicleAxles && vehicleAxleSpacings) {
+      } else if (!vehicleAxles && vehicleAxleSpacings && vehicleAxleSpacings.length > 0) {
```
Comment (gjulien-bjss) [inline src/app/features/tech-record/components/tech-record-summary/tech-record-summary.component.ts:204]:
```suggestion
      } else if (!vehicleAxles && vehicleAxleSpacings && vehicleAxleSpacings.length) {
```

File: src/app/features/tech-record/components/tech-record-summary/tech-record-summary.component.ts [hunk 6]
```diff
@@ -192,7 +192,7 @@ export class TechRecordSummaryComponent implements OnInit {
       if (vehicleAxles && vehicleAxleSpacings) {
         if (vehicleAxles.length > vehicleAxleSpacings.length + 1) {
           this.vehicleTechRecordCalculated.dimensions!.axleSpacing = this.generateAxleSpacing(vehicleAxles.length, true, vehicleAxleSpacings);
-        } else if (vehicleAxles.length < vehicleAxleSpacings.length + 1) {
+        } else if (vehicleAxles.length < vehicleAxleSpacings.length + 1 && vehicleAxleSpacings.length > 0) {
```
Comment (gjulien-bjss) [inline src/app/features/tech-record/components/tech-record-summary/tech-record-summary.component.ts:195]:
```suggestion
        } else if (vehicleAxles.length < vehicleAxleSpacings.length + 1 && vehicleAxleSpacings.length) {
```

File: src/app/forms/services/dynamic-form.service.ts (hunk 4 above)
Comment (gjulien-bjss) [reply src/app/forms/services/dynamic-form.service.ts:125]: well shows I pay attention then 😆 yep looks good then 👍

File: src/app/features/tech-record/components/tech-record-summary/tech-record-summary.component.ts (hunk 5 above)
Comment (tom-bjss) [reply src/app/features/tech-record/components/tech-record-summary/tech-record-summary.component.ts:204]: no because 0 is a length

Comment (tom-bjss) [reply src/app/features/tech-record/components/tech-record-summary/tech-record-summary.component.ts]: Take it back, code penned it, you smart

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    private actions$: Actions,
+    private globalErrorService: GlobalErrorService,
+    private route: ActivatedRoute,
+    private router: Router,
+    private store: Store<TechnicalRecordServiceState>,
+    private technicalRecordService: TechnicalRecordService
+  ) {}
+
+  ngOnInit() {
+    !this.vehicle$.subscribe(data => {
```
Comment (gjulien-bjss) [inline src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts:28]:
```suggestion
    this.vehicle$.subscribe(data => {
```
not sure what that's doing

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      if (!data) {
+        this.router.navigate(['..'], { relativeTo: this.route });
```
Comment (gjulien-bjss) [inline src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts:30]:
```suggestion
        this.navigateBack();
```

File: src/app/features/create/create.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+      component.handleSubmit();
+      tick();
+
+      expect(routerSpy).toHaveBeenCalledWith(['../create/new-record-details'], { relativeTo: route });
+    }));
+  });
+
+  describe('isVinUnique', () => {
+    it('should call isUnique with an emptry string and the type of vin', async () => {
+      const isUniqueSpy = jest.spyOn(techRecordService, 'isUnique').mockImplementation(() => of(true));
```
Comment (gjulien-bjss) [inline src/app/features/create/create.component.spec.ts:150]:
```suggestion
      const isUniqueSpy = jest.spyOn(techRecordService, 'isUnique').mockResolvedValue(true);
```
Just because the `isUnique` method returns a Promise, not an observable

Comment (gjulien-bjss) [reply src/app/features/create/create.component.spec.ts:150]: And same for the other tests in here

File: src/app/features/create/create.component.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+  get vehicleStatusOptions(): MultiOptions {
+    return [
+      { label: 'Current', value: StatusCodes.CURRENT },
+      { label: 'Provisional', value: StatusCodes.PROVISIONAL }
+    ];
+  }
+
+  get primaryVrm(): string {
+    return this.vehicle.vrms!.find(vrm => vrm.isPrimary)!.vrm;
```
Comment (gjulien-bjss) [inline src/app/features/create/create.component.ts:65]:
```suggestion
    return this.vehicle.vrms?.find(vrm => vrm.isPrimary)?.vrm || '';
```

File: src/app/features/create/create.component.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    this.router.navigate(['..'], { relativeTo: this.route });
+  }
+
+  async handleSubmit() {
+    if (!this.isFormValid || !(await this.isFormValueUnique())) return;
+
+    this.technicalRecordService.updateEditingTechRecord(this.vehicle as VehicleTechRecordModel);
+    this.technicalRecordService.generateEditingVehicleTechnicalRecordFromVehicleType(this.vehicle.techRecord![0].vehicleType);
+    await this.router.navigate(['../create/new-record-details'], { relativeTo: this.route });
+  }
```
Comment (gjulien-bjss) [inline src/app/features/create/create.component.ts:89]:
```suggestion
  handleSubmit() {
    if (!this.isFormValid || !(await this.isFormValueUnique())) return;

    this.technicalRecordService.updateEditingTechRecord(this.vehicle as VehicleTechRecordModel);
    this.technicalRecordService.generateEditingVehicleTechnicalRecordFromVehicleType(this.vehicle.techRecord![0].vehicleType);
    this.router.navigate(['../create/new-record-details'], { relativeTo: this.route });
  }
```
Not sure why you would need to await the `router.navigate`

File: src/app/app-routing.module.ts [hunk 12]
```diff
@@ -25,6 +25,12 @@ const routes: Routes = [
         canActivate: [MsalGuard, RoleGuard],
         loadChildren: () => import('./features/search/search.module').then(m => m.SearchModule)
       },
+      {
+        path: 'create',
+        data: { title: 'Create new technical record', roles: Roles.TechRecordView },
```
Comment (kpearmanbjss) [inline src/app/app-routing.module.ts:30]: Is TechRecordView the correct role for a create path ? I see we have TechRecordCreate elsewhere ?

File: src/app/services/technical-record/technical-record.service.ts (hunk 1 above)
Comment (HazelWilderspin) [reply src/app/services/technical-record/technical-record.service.ts:105]: Nick says not as part of this user flow. (TFFT)

File: src/app/features/create/create.component.ts (hunk 10 above)
Comment (HazelWilderspin) [reply src/app/features/create/create.component.ts]: fixt

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts (hunk 8 above)
Comment (HazelWilderspin) [reply src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts]: fixt

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts (hunk 7 above)
Comment (HazelWilderspin) [reply src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts]: fixt

File: src/app/features/create/create.component.spec.ts (hunk 9 above)
Comment (tom-bjss) [reply src/app/features/create/create.component.spec.ts:150]: nope

Comment (gjulien-bjss) [reply src/app/features/create/create.component.spec.ts:150]: Ignore me of course 😆

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts (hunk 3 above)
Comment (tom-bjss) [reply src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts:12]: Will update when a name comes to us but non blocking :D

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  it('get Vehicle', () => {
+    expect(component.vehicle$).toBeTruthy();
+  });
+
+  it('should navigate away on no data', () => {
+    const navigateSpy = jest.spyOn(router, 'navigate').mockImplementation(() => Promise.resolve(true));
+    jest.spyOn(component, 'vehicle$', 'get').mockImplementation(() => of(undefined));
+    component.ngOnInit();
+    expect(navigateSpy).toBeCalledWith(['..'], { relativeTo: route });
+  });
```
Comment (gjulien-bjss) [inline src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.spec.ts:57]:
```suggestion
  it('should navigate away on no data', () => {
    jest.spyOn(techRecordService, 'editableVehicleTechRecord$').mockReturnValue(of(undefined));
    const navigateSpy = jest.spyOn(component, "navigateBack")
    component.vehicle$
    expect(navigateSpy).toBeCalledTimes(1);
  });
```
I don't actually know if the suggested test passes. But this test in confusing me:
- The component does not have an onInit lifecycle hook, so I don't know why it's being called
- If you're testing behaviour from the getter, mocking said getter in the test doesn't really make sense.

Comment (tahayazici) [reply src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.spec.ts]: Done, good spot.

File: src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+      })
+    );
+  }
+
+  navigateBack() {
+    this.globalErrorService.clearErrors();
+    this.router.navigate(['..'], { relativeTo: this.route });
+  }
+
+  handleSubmit() {
```
[the remark this answers is not in the document]
Comment (tom-bjss) [reply src/app/features/create/components/hydrate-new-vehicle-record/hydrate-new-vehicle-record.component.ts:39]: Most of them are handled else where and use the global error service

**Review by HazelWilderspin — DISMISSED** [review]

**Review by kpearmanbjss — APPROVED** [review]

**Review by tahayazici — APPROVED** [review]

**Review by HazelWilderspin — APPROVED** [review]
